### PR TITLE
mds: misc fixes for MDSAuthCaps code

### PR DESCRIPTION
**Pull request by lxbsz**
Once the caller_gid is matched it will make no sense to do the following caller gid list check.

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: src/mds/MDSAuthCaps.cc [hunk 1]
```diff
@@ -145,7 +145,7 @@ bool MDSCapMatch::match(string_view target_path,
       bool gid_matched = false;
       if (std::find(gids.begin(), gids.end(), caller_gid) != gids.end())
 	gid_matched = true;
-      if (caller_gid_list) {
+      if (!gid_matched && caller_gid_list) {
```
Comment (kotreshhr) [inline src/mds/MDSAuthCaps.cc:148]: nit: This could just be `else if` statement in that case ?

**Review by kotreshhr — COMMENTED** [review]
lgtm

File: src/mds/MDSAuthCaps.cc (hunk 1 above)
Comment (lxbsz) [reply src/mds/MDSAuthCaps.cc:148]: Yeah, this also looks good to me.

File: src/mds/MDSAuthCaps.cc [hunk 2]
```diff
@@ -167,14 +167,29 @@ bool MDSCapMatch::match(string_view target_path,
...
+    if (target_path.find(_path) != 0)
       return false;
-    // if path doesn't already have a trailing /, make sure the target
-    // does so that path=/foo doesn't match target_path=/food
-    if (target_path.length() > path.length() &&
-	path[path.length()-1] != '/' &&
-	target_path[path.length()] != '/')
+    /* In case target_path.find(_path) == 0 && target_path.length() == _path.length():
+     *  path=/foo  _path=/foo target_path=/foo     --> match
+     *  path=/foo/ _path=/foo target_path=/foo/    --> match
```
Comment (kotreshhr) [inline src/mds/MDSAuthCaps.cc:181]: nit: This is the case of `target_path.length() > _path.length()`. This needs to be removed here.

Comment (lxbsz) [reply src/mds/MDSAuthCaps.cc]: Good catch. Fixed it. Thanks @kotreshhr

**Review by kotreshhr — APPROVED** [review]
lgtm

File: src/mds/MDSAuthCaps.cc
Comment (leonid-s-usov) [inline src/mds/MDSAuthCaps.cc:1]: Is there a chance you could add unit tests for the match_path function that verifies that it works correctly?

Comment (lxbsz) [reply src/mds/MDSAuthCaps.cc:1]: Yeah, makes sense. Done.

**Comment by lxbsz** [comment]
The test result:

```
[ RUN      ] AccessTest.Path
out: [{"entity":"client.libcephfs_path_test.1294787370","key":"AQA6L0xlcU7dCBAALBnE0e5CV2sAIdL6Z3X9dw==","caps":{"mds":"allow r, allow rw path=/ceph_test_libcephfs_access.good.295974217, allow rw path=/ceph_test_libcephfs_access.good_slash.1540126778/","mon":"allow r","osd":"allow rwx"}}]
key: AQA6L0xlcU7dCBAALBnE0e5CV2sAIdL6Z3X9dw==
[       OK ] AccessTest.Path (6575 ms)
```

**Review by rishabh-d-dave — APPROVED** [review]

**Review by rishabh-d-dave — APPROVED** [review]
Testing was successful - https://tracker.ceph.com/projects/cephfs/wiki/Main#16-Jan-2023

File: src/mds/MDSAuthCaps.cc [hunk 3]
```diff
@@ -167,14 +167,29 @@ bool MDSCapMatch::match(string_view target_path,
...
+
+  if (_path.length()) {
+    if (target_path.find(_path) != 0)
       return false;
-    // if path doesn't already have a trailing /, make sure the target
-    // does so that path=/foo doesn't match target_path=/food
-    if (target_path.length() > path.length() &&
-	path[path.length()-1] != '/' &&
-	target_path[path.length()] != '/')
+    /* In case target_path.find(_path) == 0 && target_path.length() == _path.length():
```
[the remark this answers is not in the document]
Comment (lxbsz) [reply src/mds/MDSAuthCaps.cc:179]: Okay, will fix this next time. Thanks @rishabh-d-dave

**Pull request merged**